### PR TITLE
Add support for slash command permissions

### PR DESCRIPTION
**Pull request by IDoEverything**
# Summary
Adds support for slash command permissions.

# Details
Added the `DefaultPermission` property and endpoints and entities for permissions.

# Changes proposed
* Add the `DefaultPermission` property to application command
* Add application command permission entities and endpoints
* Fix `GetOriginalResponse` not setting the client
* Move around interaction enums to their own folder

### CODE REVIEW

File: DSharpPlus/Net/Rest/DiscordApiClient.cs [hunk 1]
```diff
@@ -2583,17 +2589,17 @@ internal async Task CreateInteractionResponseAsync(ulong interaction_id, string
...
-            var application_id = this.Discord.CurrentApplication.Id;
             var route = $"{Endpoints.WEBHOOKS}/:application_id/:interaction_token{Endpoints.MESSAGES}{Endpoints.ORIGINAL}";
             var bucket = this.Rest.GetBucket(RestRequestMethod.GET, route, new { application_id, interaction_token }, out var path);
 
             var url = Utilities.GetApiUriFor(path);
             var res = await this.DoRequestAsync(this.Discord, bucket, url, RestRequestMethod.GET, route);
-            var json = JObject.Parse(res.Response);
+            var ret = JsonConvert.DeserializeObject<DiscordMessage>(res.Response);
 
-            return json.ToDiscordObject<DiscordMessage>();
```
Comment (VelvetThePanda) [inline DSharpPlus/Net/Rest/DiscordApiClient.cs:2596]: Why was this changed.

Comment (IDoEverything) [reply DSharpPlus/Net/Rest/DiscordApiClient.cs:2596]: Consistency with other methods

Comment (VelvetThePanda) [reply DSharpPlus/Net/Rest/DiscordApiClient.cs:2596]: Thanks for reminding me that I need to submit a PR regarding this

According to Maxine, `.ToDiscordObject` is the preferred method of deserialization

**Review by Neuheit — APPROVED** [review]
